File: libs/cua-driver/sdk/typescript/src/window.ts
/**
 * Window — a handle to a specific macOS window (pid + windowId).
 *
 * Obtain from ``App.mainWindow()`` or ``App.getWindow()``.
 */

import type { DriverClient } from './client.js'
import { Locator, type LocatorOptions } from './locator.js'

export class Window {
  constructor(
    private readonly client: DriverClient,
    public readonly pid: number,
    public readonly windowId: number,
  ) {}

  /**
   * Return a {@link Locator} for elements in this window.
   *
   * @param role  AX role to match, e.g. `"AXButton"`, `"AXTextField"`.
   * @param opts  Additional selector constraints (label, title, text, value).
   *
   * @example
   * ```ts
   * window.locator('AXButton', { label: 'OK' }).click()
   * window.locator('AXTextField').nth(1).fill('hello')
   * window.locator(undefined, { text: 'Submit' }).click()
   * ```
   */
  locator(role?: string, opts: Omit<LocatorOptions, 'role' | 'nth'> = {}): Locator {
    return new Locator(this.client, this.pid, this.windowId, { ...opts, role })
  }

  /** Capture the window as a PNG buffer. */
  async screenshot(): Promise<Buffer> {
    const result = await this.client.callTool('screenshot', {
      pid: this.pid,
      window_id: this.windowId,
    })
    const content = result['content'] as Array<{ type: string; data: string }> | undefined
    if (content) {
      for (const item of content) {
        if (item.type === 'image') return Buffer.from(item.data, 'base64')
      }
    }
    throw new Error('No image content in screenshot response')
  }

  /** Return the raw AX tree markdown, optionally filtered by query.
   *
   * ``get_window_state`` returns the AX tree as a text content block —
   * there is no structuredContent.  We scan the content array for the
   * first text item.
   */
  async getTree(query?: string): Promise<string> {
    const params: Record<string, unknown> = { pid: this.pid, window_id: this.windowId }
    if (query) params['query'] = query
    const result = await this.client.callTool('get_window_state', params)
    const content = result['content'] as Array<{ type: string; text?: string }> | undefined
    if (content) {
      for (const item of content) {
        if (item.type === 'text' && item.text != null) return item.text
      }
    }
    return ''
  }

  /** Window title from ``list_windows``. */
  async getTitle(): Promise<string> {
    const result = await this.client.callTool('list_windows', { pid: this.pid })
    const sc = (result['structuredContent'] ?? result) as Record<string, unknown>
    const windows = (sc['windows'] as Array<Record<string, unknown>> | undefined) ?? []
    for (const w of windows) {
      if (w['window_id'] === this.windowId) return (w['title'] as string | undefined) ?? ''
    }
    return ''
  }
}

File: libs/cua-driver/sdk/python/src/cua_driver/_window.py
"""Window — represents a single macOS window in a running app."""

from __future__ import annotations

import base64
from typing import TYPE_CHECKING, Optional

if TYPE_CHECKING:
    from ._client import DriverClient

from ._locator import Locator


class Window:
    """A handle to a specific window (pid + window_id pair).

    Obtain via :meth:`App.main_window` or :meth:`App.get_window`; do not
    construct directly unless you already know the window_id from
    ``list_windows``.
    """

    def __init__(
        self,
        client: "DriverClient",
        pid: int,
        window_id: int,
    ) -> None:
        self._client = client
        self._pid = pid
        self._window_id = window_id

    # ------------------------------------------------------------------
    # Locator factory
    # ------------------------------------------------------------------

    def locator(
        self,
        role: Optional[str] = None,
        *,
        label: Optional[str] = None,
        title: Optional[str] = None,
        text: Optional[str] = None,
        value: Optional[str] = None,
    ) -> Locator:
        """Return a :class:`Locator` for elements in this window.

        Parameters
        ----------
        role:
            AX role to match, e.g. ``"AXButton"``, ``"AXTextField"``.
        label:
            Matches the parenthesised description ``(label)``, ``id=label``,
            ``help="label``, or a quoted title ``"label"``.
        title:
            Matches the quoted title ``"title"`` precisely.
        text:
            Loose substring match anywhere in the AX tree line.
        value:
            Matches the ``value="..."`` attribute.

        Examples
        --------
        >>> window.locator("AXButton", label="OK").click()
        >>> window.locator("AXTextField").nth(1).fill("hello")
        >>> window.locator(text="Submit").click()
        """
        return Locator(
            self,
            role=role,
            label=label,
            title=title,
            text=text,
            value=value,
        )

    # ------------------------------------------------------------------
    # Raw access
    # ------------------------------------------------------------------

    def screenshot(self) -> bytes:
        """Capture the window and return raw PNG bytes."""
        result = self._client.call_tool("screenshot", {
            "pid": self._pid,
            "window_id": self._window_id,
        })
        content = result.get("content", [])
        for item in content:
            if item.get("type") == "image":
                return base64.b64decode(item["data"])
        raise RuntimeError("No image content in screenshot response")

    def get_tree(self, query: Optional[str] = None) -> str:
        """Return the raw AX tree markdown for this window.

        Optionally filter to lines matching *query* (passed to
        ``get_window_state``'s ``query`` parameter).
        """
        return self._get_tree(query=query)

    # ------------------------------------------------------------------
    # Properties
    # ------------------------------------------------------------------

    @property
    def title(self) -> str:
        """Window title from ``list_windows``."""
        result = self._client.call_tool("list_windows", {"pid": self._pid})
        windows = result.get("structuredContent", {}).get("windows", [])
        for w in windows:
            if w.get("window_id") == self._window_id:
                return w.get("title", "")
        return ""

    @property
    def pid(self) -> int:
        return self._pid

    @property
    def window_id(self) -> int:
        return self._window_id

    # ------------------------------------------------------------------
    # Internal
    # ------------------------------------------------------------------

    def _get_tree(self, query: Optional[str] = None) -> str:
        """Fetch AX tree markdown (used internally by Locator).

        ``get_window_state`` returns the AX tree in the text content block,
        not in structuredContent.  We scan the content array for the first
        text item.
        """
        params: dict = {"pid": self._pid, "window_id": self._window_id}
        if query:
            params["query"] = query
        result = self._client.call_tool("get_window_state", params)
        for item in result.get("content", []):
            if item.get("type") == "text":
                return item.get("text", "")
        return ""
